kyo-core/shared/src/main/scala/kyo/Resource.scala

```diff
@@ -160,10 +126,75 @@ object Resource:
         IO.Unsafe {
             val finalizer = Finalizer.Awaitable.Unsafe.init(closeParallelism)
             ContextEffect.handle(Tag[Resource], finalizer, _ => finalizer)(v)
-                .handle(IO.ensure(finalizer.close))
-                .map(result => finalizer.await.andThen(result))
+                .handle(
+                    IO.ensure(finalizer.close),
+                    Abort.run[Any]
+                ).map { result =>
+                    val io =
+                        result match
+                            case result @ Result.Error(e) =>
+                                val ex =
+                                    e match
+                                        case e: Throwable => e
+                                        case _            => new KyoException(s"Failed $e")
+                                finalizer.close(Present(ex))
+                            case _ => Kyo.unit
+                    io.andThen(finalizer.await).andThen(Abort.get(result.asInstanceOf[Result[Nothing, A]]))
```

Comment:
This code has some unsafe operations for two reasons:

1. It's necessary to handle `Abort` with `Any` to ensure all failures are caught. Without `Any`, it'd handle only panics in `Abort[Nothing]`. Given the guarantees of effect handling, it's ok to cast here since, if it's a `Failure[E]`, then there is a handler up in the stack that will handle the specific failure. I think this might be better encoded via an `Isolate` for `Abort`. I'll explore it.

3. Although `IO.ensure` now pipes the `Maybe[Throwable]`, in case of an interruption, the `IO.ensure` call will run only later when the fiber finalizes so I added special handling here. I had to make `finalizer.close` idempotent due to that.